### ISSUE TITLE
[icetray] make the segment If automatically apply to all submodules (Trac #1075)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1069]
No clue how to do this, but it's on my wishlist.

cweaver says: I guess you could get that by wrapping I3Tray.AddModule within the segment? 

Migrated from https://code.icecube.wisc.edu/ticket/1075
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:12:58", 
    "description": "No clue how to do this, but it's on my wishlist.\n\ncweaver says: I guess you could get that by wrapping I3Tray.AddModule within the segment? ", 
    "reporter": "david.schultz", 
    "cc": "jvansanten", 
    "resolution": "fixed", 
    "_ts": "1550067178841456", 
    "component": "combo core", 
    "summary": "[icetray] make the segment If automatically apply to all submodules", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2015-07-29T20:42:20", 
    "milestone": "", 
    "owner": "olivas", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20150810T23:52:04: 
* olivas changed owner from "" to "olivas"
* olivas changed status from "new" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20151201T17:28:43: 
* david.schultz commented:

> From slack:
> 
> jvs says: not hard to do, i think. icetray.traysegment just needs to become evil


* david.schultz changed cc from "" to "jvansanten"

**Comment by icecube-trac** [comment]
Trac update at 20151201T17:49:58: olivas commented:

> Another option is to create an I3MetaModule (maybe not the best name), which is a module of modules.  This was a wishlist item from long ago that never got implemented.  When segments came on the scene, they largely served the purpose that I3MetaModule was meant to.  Perhaps it's time to revisit this.
> 
> We could still have tray segments and I3MetaModules, but they would each serve different purposes depending on what you wanted to do.  One thing that would be nice, for example, is auto-magic parameter exposure.  Parameters of the contained I3Modules would naturally be parameters accessible to the I3MetaModule.

**Comment by icecube-trac** [comment]
Trac update at 20151201T18:19:00: jvansanten commented:

> An only slightly evil solution:
> 
> 
> 
> ```text
> Index: private/pybindings/I3Tray.py
> ===================================================================
> --- private/pybindings/I3Tray.py	(revision 2326)
> +++ private/pybindings/I3Tray.py	(working copy)
> @@ -51,6 +51,7 @@
>          """
>          self.last_added = None
>          self.segments_in_order = []
> +        self.current_condition = None
>          icetray.I3Tray.__init__(self)
>  
>      def _create_name(self, _type, kind, module_names):
> @@ -99,7 +100,7 @@
>          else:
>              return method(_type, _name, **kwargs)
>  
> -    def AddModule(self, _type, _name=None, **kwargs):
> +    def AddModule(self, _type, _name=None, If=None, **kwargs):
>          """
>          Add a module to the tray's processing stream.
>  
> @@ -138,6 +139,11 @@
>              self.last_added = _name
>              for k,v in kwargs.items():
>                  super(I3Tray, self).SetParameter(_name, k, v)
> +            # Handle module conditions as a special case
> +            if If is None and self.current_condition is not None:
> +                If = self.current_condition
> +            if If is not None:
> +                super(I3Tray, self).SetParameter(_name, 'If', If)
>              return self
>          except:
>              raise
> @@ -167,7 +173,7 @@
>              super(I3Tray, self).SetParameter(_name, k, v)
>          return self
>  
> -    def AddSegment(self, _segment, _name=None, **kwargs):
> +    def AddSegment(self, _segment, _name=None, If=None, **kwargs):
>          """
>          Add a tray segment to the tray. This is a small scriptlet that can
>          autoconfigure some set of modules and services according to a
> @@ -201,7 +207,11 @@
>              keys[keys.index(k)] = argnames[largnames.index(k.lower())]
>          kwargs = dict(zip(keys, vals))
>          
> -        return _segment(self, _name, **kwargs)
> +        try:
> +            self.current_condition = If
> +            return _segment(self, _name, **kwargs)
> +        finally:
> +            self.current_condition = None
>      
>      def SetParameter(self, module, param, value):
>          super(I3Tray, self).SetParameter(module, param, value)
> 
> ```
> 
> This has two features:
> 
> 1) I3Tray now explicitly knows about I3ConditionalModule. I can't decide whether that's a bad thing.
> 2) Like the existing design, it throws an error if you try to pass a condition to an unconditional module. That could be changed, which would bring convenience at the expense of foot-shooting prevention.
>

**Comment by icecube-trac** [comment]
Trac update at 20160129T20:45:04: david.schultz changed milestone from "offline V15-12-00" to "offline V16-03-00"

**Comment by icecube-trac** [comment]
Trac update at 20160318T21:18:49: 
* olivas commented:

> This is a wishlist item that's going to have to wait for the next release.


* olivas changed milestone from "offline V16-03-00" to "April 2016"

**Comment by icecube-trac** [comment]
Trac update at 20170120T19:36:31: 
* olivas changed status from "accepted" to "closed"
* olivas commented:

> In changeset 2770:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2770"
> committing jakob's proposal for conditional segment propagation to submodules.  closes #1069.
> ```


* olivas changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20170125T16:45:10: 
* david.schultz changed status from "closed" to "reopened"
* david.schultz changed resolution from "fixed" to ""
* david.schultz commented:

> On software call:
> 
> We discussed the case of segments that already handle `If` themselves.  We should detect the python function signature and if that is the case, just pass the `If` through.

**Comment by icecube-trac** [comment]
Trac update at 20170125T18:09:05: david.schultz commented:

> Hint for implementation:
> 
> ```text
> if 'If' in inspect.getargspec(tray).args:
>    # pass through the If
> else:
>    # do the new thing
> ```

**Comment by icecube-trac** [comment]
Trac update at 20170201T19:28:49: 
* david.schultz commented:

> In changeset 2781:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2781"
> allow If to pass through for segments that already define it. closes #1069
> ```


* david.schultz changed resolution from "" to "fixed"
* david.schultz changed status from "reopened" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T20:52:45: nwhitehorn commented:

> In changeset 168785:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="168785"
> committing jakob's proposal for conditional segment propagation to submodules.  closes #1069.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190111T20:53:15: nwhitehorn commented:

> In changeset 168789:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="168789"
> allow If to pass through for segments that already define it. closes #1069
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:12:58: 
* olivas commented:

> Milestone May 2016 deleted


* olivas changed milestone from "May 2016" to ""
